lib/compress/zstd_double_fast.c

```diff
@@ -396,7 +447,7 @@ size_t ZSTD_compressBlock_doubleFast_dictMatchState_generic(
                 }
             } else {
                 /* check dict long +1 match */
-                U32 const dictMatchIndexL3 = dictHashLong[dictHLNext];
+                U32 const dictMatchIndexL3 = dictHashLong[dictHLNext] >> ZSTD_SHORT_CACHE_TAG_BITS;
```

Comment:
Looks like you're not doing the tag comparison for the next position search? How come?

Reply:
I tried this and it was a couple % win for gcc11 across the board, but nuked the clang12 win. One clang12 scenario went from +45% to +30%. So I figured I'd leave it out. I can show you the numbers offline though and we can re-evaluate if you want.

Here's the commit I tested: https://github.com/embg/zstd/commit/db8033796d3b9c42251091b37bbb5cda178dd419

Reply:
I don't like that we're not using the tag here.

I understand that you saw a clang perf regression. But it just doesn't make any sense not to use a tag here. You should try again, and see if you can get clang to do better.

If you absolutely cannot, we may consider leaving the tag comparison out. But we'd need to add a comment explaining why, so the next person doesn't scratch their head wondering why we're not using the tag.

But I strongly, strongly, prefer to use the tag comparison.

Reply:
You could try not doing the dictionary load until we hit this else branch, but still using the tag.

Reply:
To clarify, the clang numbers were still very good -- just not as good. And gcc was improved across the board. So I think it's fine to just cherry-pick https://github.com/embg/zstd/commit/db8033796d3b9c42251091b37bbb5cda178dd419.